add: Multi-Z Support

BeebBeebBoob opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
<!-- Опишите, что делает ваш ПР. Документировать каждую деталь не требуется, просто укажите основные изменения. -->
Поддержка Мульти-Z. Шаг ближе к ТрамСтанции.

Описание всего:
* Добавление Multi-Z.
* Унификация файлов по атмосу в code/modules/atmospherics как если делали на других билдах(тупо надо было иначе всё разбросано к чертям)
* Обновление инициализации атмоса. По моим отчётам, (дельта только лава) смог уменьшить время с 40 секунд до 4-6 секунд обработки.
* Унификация из одного disposal.dm на несколько отдельных файлов. Сам файл имел аж 1500+ строк и там было заброшено всё в одну кучу. Кучу я поделил и разложил как должно быть.
* Минорные рефакторы аля `return 0` в `return FALSE`
* QoL: крафтинг связок. Теперь это не тупое отдельное окошко с ОДНОЙ кнопкой(!), а радиальное меню которое будет удобным.
* Обновление переменной floor_transparent разделив на три типа. Без, полу-прозрачная(пол из стекла) и полностью прозрачная. Сделано для работы пространства и кода по доступности проводки и труб и пульс демона.
* QoL: Улучшена лестница( Лестница Jacob или лестница в отеле примеры)
* Слегка перенастроен система mapping. Убран прок обработки атмоса и smooth для "грязных" загрузках. Причиной тому что сперва всё загружается, а после всё итак обрабатывается атмосом, делая во время загрузкой бесполезной.
* Переработан прок has_gravity
* Убраны дефолтные кейбинды на Home End PageUp PageDown. Заменены Look Up, Look Down, Move Up, Move Down соотвествено.
* Вероятно пофикшены спавнеры фуллтайл окон. 

## Ссылка на предложение/Причина создания ПР
<!-- Здесь оставьте ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что ваше предложение одобрено. Либо укажите, почему этот ПР должен пройти без предложки. -->
<!-- Пример ссылки: https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->
¯\_(ツ)_/¯

## Демонстрация изменений
<!-- Здесь вы можете показать изменения внешне, к примеру новые спрайты, звуки или изменения карты. Или написать, что именно изменилось для игроков. Этот пункт полностью опционален и его можно удалить. -->
![image](https://github.com/ss220-space/Paradise/assets/32985153/a0b00f33-136c-4a0d-9cbe-daf42288447f)
![image](https://github.com/ss220-space/Paradise/assets/32985153/dd2191c6-6bdb-4b1e-9520-a74e0ac090f3)
https://github.com/ss220-space/Paradise/assets/32985153/fa07e37e-3d72-40a8-b0b4-335cbcf03928 - Автор видео Den10901
![image](https://github.com/ss220-space/Paradise/assets/32985153/93de746a-d490-484d-98f2-f5843553da97)

BeebBeebBoob committed:
Big refactor: Унификация всех атмос файлов

Refactor: proc has_gravity()

BeebBeebBoob committed:
DEFINES

BeebBeebBoob committed:
Unrelated & Minor stuff in order to complete chaos

Getting crashes on trying to start the server without even getting runtimes. Instead of one giant commit I'm making on multiple for my own sane sake head.

Safe stuff because I commited those two pathfiles, duh.

multiz sheep

adding /turf/procs on zpass and zair

Updating turf-transparency defines and sanitizing abit
And finally adding openspace to be included

SAADf603 commented:
Кошмар
Я отказываюсь воспринимать данную информацию

VasilevAY commented:
Кабанчиком бы кто-то ревью 200 тыс.строчек сделал

Etrnlmelancholy commented:
Говно, переделывай

Etrnlmelancholy commented:
> Кабанчиком бы кто-то ревью 200 тыс.строчек сделал

Сделал

TrashDoxx commented:
> Кабанчиком бы кто-то ревью 200 тыс.строчек сделал

198 из которых карта.

BeebBeebBoob commented:
Мой ход? Круть. Тогда я воспользуюсь заклинанием 2 уровня "Скрытый мап в коде" используя свою 4 ячейку заклинании, что увеличивает мой код с 2к строк, до 202к строк и в итоге я смогу набрать аж до 6к8 урона. Кидаю кубики.. с преимуществом так как противник заставлен врасплох!... Охуеть 20, Крит! Ну ща посмотрим на эти сочные 12к8...

BeebBeebBoob commented on code/__HELPERS/game.dm:
```diff
@@ -37,11 +37,21 @@
+	if(istype(new_turf))
+		hand_back += new_turf
+	new_turf = get_step(center, WEST)
+	if(istype(new_turf))
```
`if(new_turf && !new_turf.density)`

BeebBeebBoob commented on code/_onclick/hud/plane_master.dm:
```diff
@@ -33,6 +33,32 @@
+/obj/screen/plane_master/transparent
+	name = "transparent plane master"
+	plane = TRANSPARENT_FLOOR_PLANE
+	appearance_flags = PLANE_MASTER
```
Эт нихрена не используется нигде. Исправить

BeebBeebBoob commented on code/controllers/subsystem/non-firing/mapping.dm:
```diff
@@ -157,7 +182,7 @@ SUBSYSTEM_DEF(mapping)
 	var/watch = start_watch()
 	log_startup_progress("Loading Lavaland...")
-	var/lavaland_z_level = GLOB.space_manager.add_new_zlevel(MINING, linkage = UNAFFECTED, traits = list(ORE_LEVEL, REACHABLE, STATION_CONTACT, HAS_WEATHER, AI_OK))
+	var/lavaland_z_level = GLOB.space_manager.add_new_zlevel(MINING, linkage = UNAFFECTED, traits = list(ORE_LEVEL, REACHABLE, STATION_CONTACT, HAS_WEATHER, AI_OK, ZTRAIT_BASETURF = /turf/simulated/floor/plating/lava/smooth/lava_land_surface))
```
Проверить Лаваленд и тайпан что они вообще загружается. Перед этим исправив рантайм на `turfs/turf.dm /Destroy`

BeebBeebBoob commented on code/controllers/subsystem/throwing.dm:
```diff
@@ -138,6 +138,10 @@ SUBSYSTEM_DEF(throwing)
 	if(callback)
 		callback.Invoke()
 
+	if(!thrownthing.currently_z_moving) // I don't think you can zfall while thrown but hey, just in case.
+		var/turf/T = get_turf(thrownthing)
+		T?.zFall(thrownthing)
```
Кидая вещи, вещица почему-то летела дальше когда "падала" вниз чем изначальный пункт назначения. Кинул на 1 клетку вниз перед собой, он летит, падает и летит вплоть до стенки.
Вероятно код падает его, меняя z уровень и чудом ломает датум броска. Датум бросок просто /shrug и говорит предмету лететь до конца.
P.S. может переместить это после окончание броска где-то?

BeebBeebBoob commented on code/datums/elements/turf_transparency.dm:
```diff
@@ -0,0 +1,73 @@
+/datum/element/turf_z_transparency
+	element_flags = ELEMENT_DETACH
+
+///This proc sets up the signals to handle updating viscontents when turfs above/below update. Handle plane and layer here too so that they don't cover other obs/turfs in Dream Maker
+/datum/element/turf_z_transparency/Attach(datum/target, is_openspace = FALSE)
```
Я хз но оно типо нормально работает?... Но оно точно не работает с floor/transparent/glass. Стекло просто тупо невидимое(хоть оно там есть).
P.S. floor/transparent/glass сам что-то шаманит с иконками? черкает свою иконку ибо старый код?

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -112,23 +114,22 @@
+ * Checks if the pulling and pulledby should be stopped because they're out of reach.
+ * If z_allowed is TRUE, the z level of the pulling will be ignored.This is to allow things to be dragged up and down stairs.
+ */
+/atom/movable/proc/check_pulling(only_pulling = FALSE, z_allowed = FALSE)
```
Это я называю прок-балласт. Он в коде не используется от слова совсем. Вся пулль хрень происходит у mob/living и это отвратительно. Не говоря что там обычный и грабные пуллы они как один целый ребёнок, только с двумя головами. Кошмар.
P.S. Пересмотреть этот прок и его потенциал??

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -34,6 +34,8 @@
 	var/thought_bubble_image = "thought_bubble"
 
+	///is the mob currently ascending or descending through z levels?
+	var/currently_z_moving
```
Я бы задебагал до чистоты эту переменную, дабы случаем код не решит "забыть" его поменять.
P.S. срать об этом в дебагер при каждом изменении? set_currently_z_moving в помощь

BeebBeebBoob commented on code/game/turfs/turf.dm:
```diff
@@ -568,6 +586,63 @@
+	return TRUE
+
+/// Precipitates a movable (plus whatever buckled to it) to lower z levels if possible and then calls zImpact()
+/turf/proc/zFall(atom/movable/falling, levels = 1, force = FALSE, falling_from_move = FALSE)
```
Целуй в зад этому проку

BeebBeebBoob commented on code/game/turfs/turf.dm:
```diff
@@ -77,6 +77,16 @@
 	if(light_power && light_range)
 		update_light()
 
+	var/turf/T = GET_TURF_ABOVE(src)
+	if(T)
+		T.multiz_turf_new(src, DOWN)
+		SEND_SIGNAL(T, COMSIG_TURF_MULTIZ_NEW, src, DOWN)
+	T = GET_TURF_BELOW(src)
+	if(T)
+		T.multiz_turf_new(src, UP)
+		SEND_SIGNAL(T, COMSIG_TURF_MULTIZ_NEW, src, UP)
```
Меня это пугает. Меня пугает что оно будет вызывать рантайм из-за попыток GET_TURF_ABOVE(src) на проверку список. 
P.S. вероятно я обосрался что не расширил multiz_level список до всех зон, дабы он не вызывал index out of list.

BeebBeebBoob commented on code/game/turfs/turf.dm:
```diff
@@ -87,6 +97,14 @@
 	if(!changing_turf)
 		stack_trace("Incorrect turf deletion")
 	changing_turf = FALSE
+
+	var/turf/V = GET_TURF_ABOVE(src)
+	if(V)
+		V.multiz_turf_del(src, DOWN)
+	V = GET_TURF_BELOW(src)
+	if(V)
+		V.multiz_turf_del(src, UP)
```
Fuck как сверху. Опять рантайм на get_turf_above в проверках списка вне границ

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -340,6 +357,116 @@
+ * * target: The target turf to move the src to. Set by can_z_move() if null.
+ * * z_move_flags: bitflags used for various checks in both this proc and can_z_move(). See __DEFINES/movement.dm.
+ */
+/atom/movable/proc/zMove(dir, turf/target, z_move_flags = ZMOVE_FLIGHT_FLAGS)
```
Этот прок буквально покрыли резиной, смазкой и другими вещами защиты, чтобы кекодеры или сам код не обосрался в попытках недотащить что либо в аргументы. Упаси.

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -340,6 +357,116 @@
+	return TRUE
+
+/// Returns a list of movables that should also be affected when src moves through zlevels, and src.
+/atom/movable/proc/get_z_move_affected(z_move_flags)
+	. = list(src)
+	if(buckled_mobs)
+		. |= buckled_mobs
+	if(!(z_move_flags & ZMOVE_INCLUDE_PULLED))
+		return
+	for(var/mob/living/buckled as anything in buckled_mobs)
+		if(buckled.pulling)
+			. |= buckled.pulling
+	if(pulling)
+		. |= pulling
```
Машины, пуллинги и другие хрени. Но есть чувство что оно не всегда работает как надо.

BeebBeebBoob commented on code/game/machinery/computer/camera_advanced.dm:
```diff
@@ -218,3 +230,31 @@
 	else
 		playsound(origin, 'sound/machines/terminal_prompt_deny.ogg', 25, 0)
+
+/datum/action/innate/camera_multiz_up
+	name = "Move up a floor"
+	button_icon = 'icons/mob/actions/actions.dmi'
+	button_icon_state = "move_up"
```
Проверить каждую камеру. Проверить что они вообще РАБОТАЮТ. Дать ответ :)

BeebBeebBoob commented on code/game/objects/effects/decals/Cleanable/misc.dm:
```diff
@@ -210,6 +210,7 @@
 	icon_state = "flour"
 	color = "#D5820B"
 	scoop_reagents = list("fungus" = 10)
+	turf_loc_check = FALSE
```
Это фунгус. Фунгусу можно на стенку, другим нельзя на стенку. :)

BeebBeebBoob commented on code/game/objects/effects/effects.dm:
```diff
@@ -103,9 +103,16 @@
 
 /obj/effect/decal/Initialize(mapload)
 	. = ..()
+	if(turf_loc_check && (!isturf(loc) || NeverShouldHaveComeHere(loc)))
+		return INITIALIZE_HINT_QDEL
```
Я бы наверное оставил бы сюда ошибку аля "Маппер. Ты обосрался. Но не волнуйся, я удалил твоё чудо в воздухе/в стенке"

BeebBeebBoob commented on code/game/objects/structures/ladders.dm:
```diff
@@ -32,19 +32,20 @@
 /obj/structure/ladder/LateInitialize()
```
Надо проверить jacob лестницу...

Den109G replied:
https://github.com/ss220-space/Paradise/assets/87372121/b56f15f0-e999-4f49-8aa8-54c9c32ba817

Лестница работает как лестница, правда пробивает всё.

BeebBeebBoob replied:
Оно должно пробивать. Это лестница против физики и логики. Мне главное чтобы оно просто работало ок с моим чудом о MultiZ

BeebBeebBoob commented on code/game/turfs/simulated/floor/transparent.dm:
```diff
@@ -20,12 +20,14 @@
-	I.plane = PLANE_SPACE
-	underlays += I
 	dir = SOUTH //dirs that are not 2/south cause smoothing jank
 	icon_state = "" //Prevents default icon appearing behind the glass
```
Может из-за этого он не хочет показывать себя...

BeebBeebBoob replied:
Дурак, ты с плейном должен был побаловаться

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -16,3 +16,186 @@
+
+/turf/simulated/openspace/Initialize(mapload)
+	. = ..()
+	if(!GET_TURF_BELOW(src))
```
Я думаю он даже не дойдёт до строчки низ и сразу крашнет инициализацию

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -16,3 +16,186 @@
+ * so flying mobs, qdeleted movables and things that were moved somewhere else during
+ * Initialize() won't fall by accident.
+ */
+/turf/simulated/openspace/proc/on_atom_created(datum/source, atom/created_atom)
```
Проверить спавн плюшек на открытом месте.

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -16,3 +16,186 @@
+		movable.set_currently_z_moving(CURRENTLY_Z_FALLING_FROM_MOVE)
+
+///Makes movables fall when forceMove()'d to this turf.
+/turf/simulated/openspace/Entered(atom/movable/movable)
+	. = ..()
+	if(movable.set_currently_z_moving(CURRENTLY_Z_FALLING))
+		zFall(movable, falling_from_move = TRUE)
```
Я вот думаю. Может спец проверку сделать, что там ничего под ним нету? Точнее дать возможность ходить в открытом пространстве по "крыше" стенки внизу. Мне напомнило айсбокс, где я мог бегать минералами которые не вскопали. 
P.S. Мы бля превращаемся в DF https://dwarffortresswiki.org/images/4/4d/DF2014_Terraform.png АХХАХАХААХ (Где ты вскапываешь только тайл, но под низ нет ещё. Хотя в DF на памяти на даёт право на такое, нооо думаю можно обмануть систему)

BeebBeebBoob commented on code/modules/mob/dead/observer/observer_movement.dm:
```diff
@@ -0,0 +1,10 @@
+/mob/dead/observer/up()
+	set name = "Move Upwards"
+	set category = "IC"
+
+	if(zMove(UP, z_move_flags = ZMOVE_FEEDBACK))
+		to_chat(src, "<span class='notice'>You move upwards.</span>")
+
+/mob/dead/observer/can_z_move(direction, turf/start, turf/destination, z_move_flags = NONE, mob/living/rider)
+	z_move_flags |= ZMOVE_IGNORE_OBSTACLES  //observers do not respect these FLOORS you speak so much of.
+	return ..()
```
Я хз но при попытках двигаться по Z, призрака посылают в нуллевую точку. 
P.S. Посмотреть как живые передвигаются, ибо у них всё нормуль

Den109G replied:
`/mob/dead/can_z_move(` просто возвращает TRUE, в это же время, у живых или атомов возвращает `destination = get_step_multiz(start, direction)`

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -340,6 +357,116 @@
+ * * z_move_flags: bitflags used for various checks. See __DEFINES/movement.dm.
+ * * rider: A living mob in control of the movable. Only non-null when a mob is riding a vehicle through z-levels.
+ */
+/atom/movable/proc/can_z_move(direction, turf/start, turf/destination, z_move_flags = ZMOVE_FLIGHT_FLAGS, mob/living/rider)
```
Добавь /obj/can_z_move() с проверками что прикручен или невозможно перемещать или может вес там + лампы падают.

BeebBeebBoob replied:
P.S. добавить доп коммент что это дичь возвращает таргет либо возвращает FALSE. Хотя там и так записано но пох

BeebBeebBoob commented on code/game/objects/structures/stairs.dm:
```diff
@@ -31,3 +31,209 @@
+			return FALSE
+	return TRUE
+
+/obj/structure/stairs_frame
```
Проверить что он нормально создаёт лестницы, а не божественное удаление турфов или проходы в нуллевую.

Den109G replied:
Оно создаёт лестницы(?) (не сжирая ресурсы), но они всегда смотрят на `2` и очень не ясно как по ним перемещаться. Хожу по ним - ничего не случается. У них есть `/datum/element/connect_loc`

https://github.com/ss220-space/Paradise/assets/87372121/057e69a7-a549-44a7-b254-e211bc7929ae

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -16,3 +16,186 @@
+		return TRUE
+	return FALSE
+
+/turf/simulated/openspace/proc/CanCoverUp()
+	return can_cover_up
+
+/turf/simulated/openspace/proc/CanBuildHere()
+	return can_build_on
+
+/turf/simulated/openspace/attackby(obj/item/C, mob/user, params)
+	if(!C || !user)
+		return TRUE
+
+	if(!CanBuildHere())
+		return
```
Мне надо реально решить проблемы с постройками на открытом пространстве. Ну или нахрен удалить CanCoverUp и CanBuildHere...

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -16,3 +16,186 @@
+				new /obj/structure/lattice/catwalk/fireproof(src)
+	..()
+
+// PARACODE
```
Это часть нашего парадизовского которое я отобрал у simulated/floor. Не орите

BeebBeebBoob commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -86,8 +89,8 @@ GLOBAL_DATUM_INIT(_preloader, /datum/dmm_suite/preloader, new())
 					if(zcrd > world.maxz)
-						if(shouldCropMap)
-							continue
+						if(!increaseZ)
+							continue // Beware of runtimes. Beware of trying to add content on non-existed z level.
```
Если чудом код достиг этого пункта, то надеюсь что у нас есть готовые пустые уровни иначе готовься к спаму рантаймом о недоступности

BeebBeebBoob commented on code/modules/mob/living/living.dm:
```diff
@@ -79,6 +79,64 @@
+		if(leg.has_fracture())
+			functional_legs = FALSE
+			break
+	if(((istajaran(src) && functional_legs) || cat) && !lying && can_help_themselves)
```
по поводу таяр я как-то sus.... Хотя если они крупные, то пусть взамен стана просто получат побольше урона, мини-коты и животные-коты не получат такого прикола.

Den109G replied:
Таярам дай и больше урона и больше стана. Они космо-коты, они не знают как приЗЕМЛЯтся.

BeebBeebBoob replied:
P.S. ломает только колени, но и ботинки(foot)

BeebBeebBoob commented on code/modules/mob/living/living.dm:
```diff
@@ -627,19 +685,18 @@
 
 	return
-
+currently_z_moving
```
FUCK

BeebBeebBoob commented on code/modules/mob/living/living.dm:
```diff
@@ -666,6 +723,10 @@
 		var/pull_dir = get_dir(src, pulling)
 		pulling.glide_size = glide_size
 		if(get_dist(src, pulling) > 1 || (moving_diagonally != SECOND_DIAG_STEP && ((pull_dir - 1) & pull_dir))) // puller and pullee more than one tile away or in diagonal position
+			// This sucks.
+			// Pulling things up/down & into other z-levels. Conga line lives.
+			if(pulling.z != z)
+				dest = get_step(pulling, get_dir(pulling, dest))
```
Всё ещё сосёт. Я вот реально должен сломать себе голову чтобы оно заработало.

BeebBeebBoob replied:
Дружбан прошлый постой. Есть три способа в которых pulling надо передвинуть по moveZ
1. Тот кто тащит перемещается через Move Up или Move Down 
2. Тот кто тащит использовал /ladder
3. Тот кто тащит переместился через /stairs
Может просто перехватывать и дотаскивать в тех местах?

BeebBeebBoob commented on code/modules/mob/living/silicon/ai/freelook/eye.dm:
```diff
@@ -142,6 +151,21 @@
 	acceleration = !acceleration
 	to_chat(usr, "Camera acceleration has been toggled [acceleration ? "on" : "off"].")
 
+/mob/living/silicon/ai/up()
+	set name = "Move Upwards"
+	set category = "IC"
+
+	if(eyeobj.zMove(UP, z_move_flags = ZMOVE_FEEDBACK))
+		to_chat(src, span_notice("You move upwards."))
+
+/mob/living/silicon/ai/down()
+	set name = "Move Down"
+	set category = "IC"
+
+	if(eyeobj.zMove(DOWN, z_move_flags = ZMOVE_FEEDBACK))
+		to_chat(src, span_notice("You move down."))
```
Проверить.

Den109G replied:
Работает?

[Ссылка на видео.](https://discord.com/channels/617003227182792704/734823601110515882/1196686633546829975) В конце не мог вверх-вниз управлять камерой. Рантаймов нет. Гравитация действует на глаз ИИ.

BeebBeebBoob commented on code/modules/power/supermatter/supermatter.dm:
```diff
@@ -443,6 +443,11 @@
 
 	Consume(moving_atom)
 
+/obj/machinery/power/supermatter_crystal/intercept_zImpact(list/falling_movables, levels)
+	. = ..()
+	for(var/atom/movable/hit_object as anything in falling_movables)
+		Bumped(hit_object)
+	. |= FALL_STOP_INTERCEPTING | FALL_INTERCEPTED
```
Оно не сработало когда я упал прям на СМ.

BeebBeebBoob commented on code/modules/surgery/organs/blood.dm:
```diff
@@ -259,6 +259,8 @@
 		return
 	if(!T)
 		T = get_turf(src)
+	if(isopenspaceturf(T) && !GET_TURF_BELOW(T))
+		return
```
Это надо кстати допилить и в других местах, чтобы не было приколов аля "кровь левитирует".

BeebBeebBoob reviewed (commented):
XP

BeebBeebBoob commented:
Был конфликт: [refactor: Icon Updates Rework Part1](https://github.com/ss220-space/Paradise/commit/fd2a97e295bd9152139ddd30bbef821e697cb1f2)
Решён

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -368,6 +385,116 @@
+			else
+				to_chat(src, "<span class='notice'>You are not Superman.<span>")
+		return FALSE
+	if(!(z_move_flags & ZMOVE_IGNORE_OBSTACLES) && !(start.zPassOut(src, direction, destination) && destination.zPassIn(src, direction, start)))
```
`(!(z_move_flags & ZMOVE_ALLOW_ANCHORED) && anchored))`

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -0,0 +1,201 @@
+/turf/simulated/openspace/zAirOut()
+	return TRUE
+
+/turf/simulated/openspace/zPassIn(atom/movable/A, direction, turf/source)
+	if(direction == DOWN)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_IN_DOWN)
+				return FALSE
+		return TRUE
+	if(direction == UP)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_IN_UP)
+				return FALSE
+		return TRUE
+	return FALSE
```
```suggestion
/turf/simulated/openspace/zPassIn(direction)
	if(direction == DOWN)
		for(var/obj/contained_object in contents)
			if(contained_object.obj_flags & BLOCK_Z_IN_DOWN)
				return FALSE
		return TRUE
	if(direction == UP)
		for(var/obj/contained_object in contents)
			if(contained_object.obj_flags & BLOCK_Z_IN_UP)
				return FALSE
		return TRUE
	return FALSE
```

BeebBeebBoob commented on code/game/turfs/simulated/openspace.dm:
```diff
@@ -0,0 +1,201 @@
+		return TRUE
+	return FALSE
+
+/turf/simulated/openspace/zPassOut(atom/movable/A, direction, turf/destination)
+	if(direction == DOWN)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_OUT_DOWN)
+				return FALSE
+		return TRUE
+	if(direction == UP)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_OUT_UP)
+				return FALSE
+		return TRUE
+	return FALSE
```
```suggestion
/turf/simulated/openspace/zPassOut(direction)
	if(direction == DOWN)
		for(var/obj/contained_object in contents)
			if(contained_object.obj_flags & BLOCK_Z_OUT_DOWN)
				return FALSE
		return TRUE
	if(direction == UP)
		for(var/obj/contained_object in contents)
			if(contained_object.obj_flags & BLOCK_Z_OUT_UP)
				return FALSE
		return TRUE
	return FALSE
```

BeebBeebBoob commented on code/game/turfs/turf.dm:
```diff
@@ -540,6 +558,88 @@
 /turf/AllowDrop()
 	return TRUE
 
+/turf/proc/zPassIn(atom/movable/A, direction, turf/source)
+	if(direction == DOWN)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_IN_DOWN)
+				return FALSE
+		return TRUE
+	return FALSE
+
+//direction is direction of travel of A
+/turf/proc/zPassOut(atom/movable/A, direction, turf/destination)
+	if(direction == UP)
+		for(var/obj/O in contents)
+			if(O.obj_flags & BLOCK_Z_OUT_UP)
+				return FALSE
+		return TRUE
+	return FALSE
```
добавить проверку на density, мы не хотим приколов работая со стенками, путая их с полом

BeebBeebBoob replied:
```suggestion
//The zpass procs exist to be overriden, not directly called
//use can_z_pass for that
///If we'd allow anything to travel into us
/turf/proc/zPassIn(direction)
	return FALSE

///If we'd allow anything to travel out of us
/turf/proc/zPassOut(direction)
	return FALSE
```

А проверки засунуть в /turf/simulated с density
Не забудь также для отдельного тупого turf/space прописать как прописано с openspace

BeebBeebBoob reviewed (commented):
Больше комментов для себя

BeebBeebBoob committed:
Fixes and Atmos 2: Air Boogaloo

BeebBeebBoob commented:
+3797 -1838 строк... Найс. Остался Атмос, ещё полно фиксов, несколько иконок и я думаю можно пускать людей в пропасть.
БашняСтанция, я иду!

BeebBeebBoob commented:
Блин! Моё царство нарушено! Будь проклят этот мерж конфликт!

TrashDoxx commented:
> Блин! Моё царство нарушено! Будь проклят этот мерж конфликт!

Жесть ты шизик. Вот что хоумстак с людьми делает.

BeebBeebBoob committed:
The Finale: Multi-Z Support

Это должен быть последний коммит, который выводит этот ПР на уровень возможности для мержа. Было пролито много пота, что уже переливает ведро.

Linters on map

Linters cable_coil

BeebBeebBoob commented:
Готово. Если будут конфликты с другими ПРами, решу без проблем.

BeebBeebBoob committed:
Look up & Down, Keybinds and bugfixes

Это вот прям последий коммит. Дальше будут обновления и багфиксы потенциальные.

BeebBeebBoob committed:
Stairs to build & Now you can multi-z explode

"forgets logs" Oh no! Anyway.

Fix linter

PhantornRU commented:
Психопат! В хорошем смысле

BeebBeebBoob commented:
Rebase это кайф.

BeebBeebBoob commented:
Я слегка обосрался и оставил кусочки моей новой карты здесь. Оно всё равно не будет в ротации, но зато админам будет весело на нём(может).

Den109G commented:
dear God, please add sound effect as mob falls through grille it would be so funny (making a thread with videos of play-testing)

https://github.com/ss220-space/Paradise/assets/87372121/2efbc2f8-7c45-43d6-88e9-554e6c8c0837

Den109G commented:
some weird rabbit-jump physics

https://github.com/ss220-space/Paradise/assets/87372121/1b4b4f52-2043-41c5-b44c-f334dcc1aede

Den109G commented:
Новая страта как искать фальш стены, ценой своих ног.

https://github.com/ss220-space/Paradise/assets/87372121/df7f23f9-a1ff-4db8-85b9-aaadf55287ae

Den109G commented:
Призраки не Супермены. Верно. Но, по моему скромному знанию паранормальных явлений, они должны мочь лететь вниз/через пол.
(Проверял как газ вообще распространяется. Вроде бы нормально)

https://github.com/ss220-space/Paradise/assets/87372121/a20f2035-e5aa-4f7f-bcc8-c8c05cc2cfd0

Den109G commented:
Телекинез кидает вещи дальше, если объект на другом уровне.

https://github.com/ss220-space/Paradise/assets/87372121/1365627b-496a-4355-b9db-f4becab9cd1e

Телекинез может.... C другого этажа ударить по челу на том этаже, где стоит?

https://github.com/ss220-space/Paradise/assets/87372121/e2690122-40e1-42ee-a890-e190cae2ef36

https://github.com/ss220-space/Paradise/assets/87372121/afd669ca-edb5-43d6-b0b6-86f804fa817c

Den109G commented:
WANTED: tajaran "Long Paws"
DEAD OR ALIVE
Stole a cap's spare, while he was drunk

https://github.com/ss220-space/Paradise/assets/87372121/a62fb71a-b103-4ac3-b353-448d37c2a5c9

Den109G commented:
~~Движок звуков решил выйти за грань нашего понимания. В теории, может софтлокать игру.
Я это только встречал на Мульти-З. На основной ветке - такого ещё не случалось. Где-то каждые 20 минут ломается, при активном пользовании мульти-З~~
В 2568 году учёные доказали, что кроме двух осей координат X и Y есть ещё и третья координата - W. И так, компания NanoTrahen Energy on plasma изобрела катушки, что порабощает людей, используя инфракрасные звуки. И так в мире сломались все уши.

https://github.com/ss220-space/Paradise/assets/87372121/5c102870-a991-409f-8853-ed468c1bafa7

BeebBeebBoob commented:
+ от @Den109G из дискорда:
хелп/харм интент на лестницу пофиксить надо, (if harm do this)
пофиксить падение при обновлении obj_flags(точнее при разборе палок)

Уже проблемы(выше по видео):
Урон при падении на грилль? (ZMOVE_INTERCEPT)
Ген прыжка (Забрать flying на последний шаг, дабы турф openspace правильно обработал zFall() )
Фальш стены дать obj_flags ( z move in и z move out, уже на моём компе сделано)
Призраки должны летать также вниз (обосрался при коде. Говно убрал)
Вообще Нахрен Перелопатить Взаимодействие с турфами ниже. Особенно телекинез и возможность брать вещи с помощью Look Up и Look Down. (Обновить проверку на Is Adjecent для саппорта Multi Z)

BeebBeebBoob commented:
Себе:
[сюда](https://github.com/ss220-space/Paradise/blob/master/code/_onclick/adjacent.dm#L34)
```
	if(get_dist(src, T0) > 1 || z != T0.z) //too far
		return FALSE
```

Den109G commented:
> Урон при падении на грилль? (ZMOVE_INTERCEPT)

Просто та же причина что и у фальш стены. Ходишь по стене - и вдруг бац, падаешь между прутей. Как к видео оставил приписку: со звуком было бы смешней.

BeebBeebBoob committed:
NUCLEAR PATCHES INCOMING

BeebBeebBoob committed:
tweaking jump gene and lattice fuckup

LINTER

BeebBeebBoob committed:
no more transparent in glass + Multi-z Explosion tweak and linters

WINDOW FIX. FUCK.

BeebBeebBoob commented:
Todo: пофиксить блюр

BeebBeebBoob committed:
replanes to fix blurry effects on openspace

TrashDoxx commented:
Давай-давай, пили мульти-левел, нам Церу маппить надо.

BeebBeebBoob committed:
Multi-z support Nanomap, space air fix and other fixes and tweaks

Delete dmm-tools.exe

Update tgui.bundle.js

BeebBeebBoob commented:
Уходит в драфт. Мне не нравиться инициализация атмоса.

BeebBeebBoob committed:
Atmos refined

BeebBeebBoob commented:
Перевернул атмос. Атмосу понравилось.... 😊 Снова готов для мержа.

BeebBeebBoob committed:
Atmos Double Fixed. Fucked spacehotel. also sound fix. Chasm fix.

BeebBeebBoob committed:
Pathfinding fix on openspace

linter and stop making procs with upper latters. only lowercase and underscores WTF

BeebBeebBoob committed:
Linter delta fix

BeebBeebBoob commented:
Бля. Ну... "Conflicting files tgui/packages/tgui/public/tgui.bundle.js" ...
Я эт не могу исправить сорянб :clueless:

BeebBeebBoob commented on code/game/objects/explosion.dm:
```diff
@@ -138,8 +144,25 @@
+		if(multiz_explosions)
+			var/turf/above = GET_TURF_ABOVE(epicenter)
+			var/turf/below = GET_TURF_BELOW(epicenter)
+			floor_block[z0] = epicenter.explosion_vertical_block
```
Понять, я правильно чёрт возьми нормально сделал взрыв Multi-Z или нет. Последний раз когда я проверял то оно мне выдавало рантаймы на эту строку. ХМММ

BeebBeebBoob replied:
Себе также. Подправить isinspace() прок

BeebBeebBoob committed:
Fix: explosion, tgui bundle. Add: throwing E on active_start_turfs

And fix spacehotel

BeebBeebBoob committed:
Some touches

BeebBeebBoob commented:
WHAT THE FUCK>>>>>>>>>>>

BeebBeebBoob commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -470,7 +470,9 @@
 
 /obj/structure/table/glass/proc/check_break(mob/living/M)
-	if(has_gravity(M) && M.mob_size > MOB_SIZE_SMALL)
+	if(M.flying || M.incorporeal_move)
+		return
+	if(M.has_gravity() && M.mob_size > MOB_SIZE_SMALL)
```
Uuuh Fuck

BeebBeebBoob commented on code/game/atoms_movable.dm:
```diff
@@ -368,6 +397,110 @@
+			if(z_move_flags & ZMOVE_FEEDBACK)
+				to_chat(rider || src, "<span class='warning'>There's nowhere to go in that direction!</span>")
+			return FALSE
+	var/mob/living/L = src
+	var/flyoting = FALSE
+	if(istype(L))
+		flyoting = L.flying || L.floating // One is from /mob, the OTHER goes from /mob/living. WTF
+	if(z_move_flags & ZMOVE_FALL_CHECKS && (throwing || flyoting || !has_gravity(start)))
+		return FALSE
+	if((z_move_flags & ZMOVE_CAN_FLY_CHECKS) && !flyoting && has_gravity(start))
```
Обновить после #4249 
```suggestion
	if(z_move_flags & ZMOVE_FALL_CHECKS && (throwing || (movement_type & (FLYING|FLOATING)) || !has_gravity(start)))
		return FALSE
	if(z_move_flags & ZMOVE_CAN_FLY_CHECKS && !(movement_type & (FLYING|FLOATING)) && has_gravity(start))
```

BeebBeebBoob committed:
eeehh.. Fuck it. ATmos fix and improvement

Cool, atmos gets better and better

BeebBeebBoob commented:
ПР не запушится сама! ТОЛКАЕМ

Etrnlmelancholy commented:
Заводи жигуль, а не болтай! Мы толкаем пока

ArtemDerp commented:
/me тужится, пыжится и толкает мултижигуль

SAADf603 commented on code/modules/mob/mob_movement.dm:
```diff
@@ -511,3 +511,57 @@
+		if(buckled)
+			to_chat(src, "<span class='notice'>[buckled] is is not capable of flight.<span>")
+		else
+			to_chat(src, "<span class='notice'>You are not Superman.<span>")
```
```suggestion
			to_chat(src, "<span class='warning'>There's nowhere to go in that direction!</span>"))
```

BeebBeebBoob replied:
You are not Superman - Это когда сверху есть пространство но гравитация махает пальцем. Ты не супермен. Ты не умеешь летать.

SAADf603 commented on code/game/atoms_movable.dm:
```diff
@@ -367,6 +396,110 @@
+			if(rider)
+				to_chat(rider, "<span class='notice'>[src] is is not capable of flight.<span>")
+			else
+				to_chat(src, "<span class='notice'>You are not Superman.<span>")
```
```suggestion
				to_chat(src, "<span class='warning'>There's nowhere to go in that direction!</span>")
```

BeebBeebBoob replied:
Пару строк выше: to_chat(rider, "<span class='notice'>[src] is is not capable of flight.<span>")
:D

SAADf603 reviewed (changes requested):
На данный момент это единственные замечания. Фраза "ты не супермен" - выглядит достаточно неуместно к попыткам попасть на уровень ниже.
